File: javascript/219.contains-duplicate-ii.js
/*
 * @lc app=leetcode id=219 lang=javascript
 *
 * [219] Contains Duplicate II
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var containsNearbyDuplicate = function (nums, k) {
  // Solution 1:
  // const appear = {};
  // for (let i = 0; i < nums.length; i++) {
  //   const num = nums[i];
  //   if (appear.hasOwnProperty(num) && i - appear[num] <= k) return true;
  //   appear[num] = i;
  // }

  // Solution 2:
  for (let i = 0; i < nums.length; i++) {
    const n = nums.indexOf(nums[i], i + 1);
    if (n != -1 && n - i <= k) return true;
  }

  return false;
};
// @lc code=end

File: rust/219.contains-duplicate-ii.rs
/*
 * @lc app=leetcode id=219 lang=rust
 *
 * [219] Contains Duplicate II
 */

// @lc code=start
impl Solution {
    pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
        use std::collections::HashMap;
        let mut index_map: HashMap<i32, usize> = HashMap::with_capacity(nums.len());
        for i in 0..nums.len() {
            if let Some(index) = index_map.insert(nums[i], i) {
                if (i - index) as i32 <= k {
                    return true;
                }
            }
        }
        return false;
    }
}
// @lc code=end
